### ISSUE TITLE
Vim: Home and End keys not working since GfW 2.27.0

### ISSUE THREAD

**Issue by jberezanski** [git-for-windows/git#2854]
- [x] I was not able to find an [open](https://github.com/git-for-windows/git/issues?q=is%3Aopen) or [closed](https://github.com/git-for-windows/git/issues?q=is%3Aclosed) issue matching what I'm seeing

### Setup

 - Which version of Git for Windows are you using? Is it 32-bit or 64-bit?

```
$ git --version --build-options

git version 2.28.0.windows.1
cpu: x86_64
built from commit: 77982caf269b7ee713a76da2bcf260c34d3bf7a7
sizeof-long: 4
sizeof-size_t: 8
shell-path: /bin/sh
```

 - Which version of Windows are you running? Vista, 7, 8, 10? Is it 32-bit or 64-bit?

64-bit Windows 10 1909 Home, Polish language

```
$ cmd.exe /c ver

Microsoft Windows [Version 10.0.18363.1139]
```

 - What options did you set as part of the installation? Or did you choose the
   defaults?

Initial install and subsequent upgrades done via Chocolatey, with default options.

```
> type "C:\Program Files\Git\etc\install-options.txt"

Editor Option: VIM
Custom Editor Path:
Path Option: Cmd
SSH Option: OpenSSH
Tortoise Option: false
CURL Option: OpenSSL
CRLF Option: CRLFAlways
Bash Terminal Option: MinTTY
Performance Tweaks FSCache: Enabled
Use Credential Manager: Enabled
Enable Symlinks: Disabled
```

 - Any other interesting things about your environment that might be related
   to the issue you're seeing?

None I can think of.

### Details

 - Which terminal/shell are you running Git from? e.g Bash/CMD/PowerShell/other

Depending on the situation, one of these:
PowerShell 5.1
CMD
PowerShell 5.1 started from CMD

 - What commands did you run to trigger this issue? If you can provide a
   [Minimal, Complete, and Verifiable example](http://stackoverflow.com/help/mcve)
   this will help us understand the issue.

In any repository:
```
git commit --amend
```

Then attempt to navigate Vim using Home and End keys.

 - What did you expect to occur after running these commands?

Home key should move the text cursor to the start of the current line.
End key should move the text cursor to the end of the current line.

 - What actually happened instead?

Home key causes `^[[H` to appear briefly at the bottom of Vim window and the text cursor is not moved.
End key causes `^[[F` to appear briefly at the bottom of Vim window and the text cursor is not moved.

The issue appeared when I upgraded Git for Windows from 2.25.0 (which I had been using for some time) to 2.28.0.
After discovering the issue, I downgraded back to 2.25.0 and installed each subsequent release one by one. That way I was able to determine that the issue first appeared in 2.27.0 (2.26.2 was the last properly working version).

From reading other issues I picked up the idea of running `:version` in Vim, but the output is exactly the same between GfW 2.26.2 and 2.27.0, so no clue there.

I don't really know anything about configuring vim or diagnosing such issues, so please advise if there is anything I can check or try.

**Comment by dscho** [comment]
Please test the latest [snapshot](https://wingit.blob.core.windows.net/files/index.html).

**Comment by jberezanski** [comment]
Installed

Thu, 15 Oct 2020 22:32:55 +0200
(commit e2ca5f02f09b717ccc60f796df0404877b5443d5)
git version 2.29.0.rc2.windows.1

but no change in Vim behavior, unfortunately.

**Comment by dscho** [comment]
I just verified this myself, and came up with a workaround: `$env:TERM = "xterm-256color"` (it also works with `"xterm"`).

Now, how to fix this properly? I don't know... We could, technically, set `TERM` in the [Git wrapper](https://github.com/git-for-windows/git/wiki/The-%22Git-wrapper%22). However, the `TERM` variable is supposed to describe the current terminal, and it is a bit iffy if Git forces a value there...

**Comment by jberezanski** [comment]
I don't have this variable set on any of my machines. I suppose this means that somewhere between my shell and vim.exe something sets `TERM` to some default value. The process tree visible in Process Explorer is `cmd.exe` -> `git.exe` (wrapper) -> `git.exe` (real) -> `sh.exe` -> `vim.exe`. Now I wonder - what has changed between 2.26.2 and 2.27.0 to cause this issue? Did the logic for determining the default `TERM` value change...?

On a more practical side: how about setting `TERM` in the wrapper, but only if it is not set already?

And yes, `xterm` or `xterm-256color` do fix `Home` and `End` keys, but they also change colors in Vim to a darker and less readable palette, which is rather unfortunate.
without TERM:
![image](https://user-images.githubusercontent.com/1577094/96320045-82cea000-1011-11eb-849b-ce893c2414fc.png)
with TERM set to xterm:
![image](https://user-images.githubusercontent.com/1577094/96320118-bad5e300-1011-11eb-94a6-4ff4c33f4f47.png)
with TERM set to xterm-256color:
![image](https://user-images.githubusercontent.com/1577094/96320164-e953be00-1011-11eb-83b3-b8648c5ee96b.png)

One additional observation: with `TERM` unset, after exiting Vim, the console screen is not cleared. All the text shown by Vim remains on the screen and the command prompt is printed over it (starting from the top of the screen). This, too, started happening in 2.27.0 - previously the screen was cleared properly after exiting Vim. Setting `TERM` fixes this.

**Comment by jberezanski** [comment]
Spelunking inside the Git for Windows install directory, I found `usr\share\vim\vim82\doc\term.txt`. It inspired me to examine the `term` setting inside Vim (with `TERM` environment variable unset), which was `cygwin`:
```
:set term
  term=cygwin
```
When I set it to xterm (`:set term=xterm`), interesting things happened:
- the color palette remained unchanged (the more readable, brighter one),
- Home and End started working,
- the screen was not cleared after exiting Vim.

On the other hand, setting `TERM` to `cygwin` results in the same behavior as with `TERM` unset. I suppose this means that `cygwin` is the default `TERM` value and its handling changed somehow in 2.27.0 - possibly with the upgrade of the MSYS runtime.

**Comment by dscho** [comment]
Oh... maybe we suddenly forget the `cygwin` termcap now?

**Comment by dscho** [comment]
Hrm. We do seem to ship it (it's in `/usr/share/terminfo/63/cygwin`)...

**Comment by dscho** [comment]
On a hunch, I just bisected this to a change of the `usr\bin\msys-2.0.dll` file between git-for-windows/git-sdk-64@4114e034d6 and git-for-windows/git-sdk-64@dfeff18c6f. Sadly, this means that the (big!) jump from MSYS2 runtime v3.0.7 to v3.1.4 brought in that regression.

**Comment by dscho** [comment]
My current hunch is that git-for-windows/msys2-runtime@bd627864ab4189984cdb0892c00f91e39c4e8243 should have changed the fallback from `cygwin` to `xterm`.

**Comment by dscho** [comment]
I'll have to stop my investigation at this point for now.

**Comment by TaleTN** [comment]
FYI: If you enable the "Use legacy console" checkbox on the Options tab of the Windows 10 Command Prompt, then this fixes this issue, at least it does at my end (on Win10 Pro x64 1909).

Do note that this isn't a solution (not for me anyway), because this system-wide setting causes other issues, e.g. you can't use Ctrl-C/V to copy/paste anymore.

**Comment by TaleTN** [comment]
Some extra info: If I install 2.29.0, but then manually copy msys-2.0.dll from 2.26.2 to %ProgramFiles%\Git\usr\bin, then everything seems fine.

**Comment by TaleTN** [comment]
FYI: This is still an issue with 2.30.0. By now adding this to the known issue list in the release notes might be in order (IMHO). Anyway, rolling back to 2.26.2, again...

**Comment by sharpau** [comment]
I also have this issue on git for windows 20.30.2.

**Comment by dscho** [comment]
> 
> 
> My current hunch is that [git-for-windows/msys2-runtime@bd62786](https://github.com/git-for-windows/msys2-runtime/commit/bd627864ab4189984cdb0892c00f91e39c4e8243) should have changed the fallback from `cygwin` to `xterm`.

For the record, this can be implemented by _any_ interested contributor. Opening a PR at https://github.com/git-for-windows/msys2-runtime will trigger a workflow run that builds the MSYS2 runtime and publishes it as an artifact. This artifact can be downloaded and tested. Therefore, the amount of effort required to try to fix this is really small.

**Comment by TaleTN** [comment]
I'm sorry, but I'm not really sure what exactly you mean... Do note that, like I already said a few months back, set term=xterm doesn't fully fix the issue for me.

**Comment by dscho** [comment]
> I'm sorry, but I'm not really sure what exactly you mean..

@TaleTN https://github.com/git-for-windows/msys2-runtime/commit/bd627864ab4189984cdb0892c00f91e39c4e8243 changed the fallback from `TERM=cygwin` to `TERM=xterm256-color`: https://github.com/git-for-windows/msys2-runtime/commit/bd627864ab4189984cdb0892c00f91e39c4e8243#diff-58d9d402d5db25b56c1a6d36aaa6b8af563f298144a471eaa82b23053bb96580R906

To fix this, you would have to edit https://github.com/git-for-windows/msys2-runtime/commit/bd627864ab4189984cdb0892c00f91e39c4e8243#diff-58d9d402d5db25b56c1a6d36aaa6b8af563f298144a471eaa82b23053bb96580R873 to say `xterm` instead of `xterm256-color`, then commit that change, then push it to your own fork of `msys2-runtime`. Make sure that Actions are enabled in that fork. Then wait until the build is finished, download the build artifact, extract `msys-2.0.dll` from there, replace your local Git for Windows installation's version with it, and test (but do undo your `set term=xterm` workaround).

If all works, open a PR and we'll merge it.

It is not complicated at all to do all these steps, it just takes time, and since you want this, you should have all the incentive you need to get going.

**Comment by dscho** [comment]
> you would have to edit [git-for-windows/msys2-runtime@bd62786#diff-58d9d402d5db25b56c1a6d36aaa6b8af563f298144a471eaa82b23053bb96580R873](https://github.com/git-for-windows/msys2-runtime/commit/bd627864ab4189984cdb0892c00f91e39c4e8243#diff-58d9d402d5db25b56c1a6d36aaa6b8af563f298144a471eaa82b23053bb96580R873) to say `xterm` instead of `xterm256-color`, then commit that change, then push it to your own fork of `msys2-runtime`.

This is the precise location of the line you need to edit: https://github.com/git-for-windows/msys2-runtime/blob/12e88192b84673b81c6550e48fb7df9b607dada3/winsup/cygwin/environ.cc#L915.

You do not even need to clone the repository, just open https://github.dev/git-for-windows/msys2-runtime/blob/HEAD/winsup/cygwin/environ.cc#L915 and work on it in a virtual file system (where you can commit directly into your fork on GitHub).

**Comment by TaleTN** [comment]
@dsco Thanks for explaining! I first pushed the change, and then enabled Actions, which was the wrong way around I guess, oops. Anyway, after pushing a revised commit it started the build, and it has since finished it successful, woohoo! However, there don't appear to be any artifacts (it just says "–"), so I guess I must have done something wrong. Any pointers?

**Comment by dscho** [comment]
Oh... could you cherry-pick https://github.com/msys2/msys2-runtime/commit/fa13d627f515aeee14cd256ba658a02751330dbd#diff-87ee5504a3e25ac558b343724c905f2f7949e8cec3d92b9c4300bb922afa164f ? I missed that Git for Windows' fork does not yet have that commit.

**Comment by TaleTN** [comment]
Yeah, that looks like what I've just added, except I haven't cherry-picked yours (but hopefully that won't matter). Thanks!

**Comment by dscho** [comment]
Since it is technically _not_ a fork (at least not as far as the commit history goes), it might be easier to download https://github.com/msys2/msys2-runtime/commit/fa13d627f515aeee14cd256ba658a02751330dbd.patch and to `git am` it.

**Comment by TaleTN** [comment]
For now I've already manually just added what I thought was missing when comparing my build.yaml with one of yours from a while back. I guess if it actually works, and if the change actually fixes the issue, then maybe I should do it again properly before doing a PR?

**Comment by dscho** [comment]
Please first open the PR if everything works as hoped. I can always help (when I'm not "taking a day off", as I originally intended to do today :smile:).

**Comment by TaleTN** [comment]
Well, I was able to download the artifact, but it didn't fix the issue. Looking more closely at commit bd627864ab I see that has_con_24bit_colors changes not only TERM, but some other stuff as well, so for testing purposes I'm now going to test if disabling has_con_24bit_colors has any effect.

**Comment by TaleTN** [comment]
Oh yeah, that works, issue fixed! However, my "fix" is bound to make someone somewhere unhappy (assuming has_con_24bit_colors was added for a reason). Shall I open a PR nevertheless?

**Comment by TaleTN** [comment]
PR opened, but I forgot to sign-off, so I guess I have to fix that...

**Comment by dscho** [comment]
> However, my "fix" is bound to make someone somewhere unhappy (assuming has_con_24bit_colors was added for a reason).

That sounds like a reasonable assumption.

The biggest question is: what of those `has_con_24bit_colors` things is/are causing the problem? Can we fix it?

**Comment by TaleTN** [comment]
Another solution might be to somehow detect whether the user has legacy console for the command prompt enabled or not. If legacy mode is disabled, then this should override override has_con_24bit_colors and set it to 0. Apparently there is a HKCU registry key for this, so this should be possible.

**Comment by dscho** [comment]
> Another solution might be to somehow detect whether the user has legacy console for the command prompt enabled or not.

Hmm. I don't really know whether we _should_. I would like to understand first what the root cause of those problems is. Do we need to use a different set of termcap settings, depending whether the legacy console is enabled or not? To decide that, we would need to know what _precisely_ is going wrong under `has_con_24bit_colors`.

**Comment by rimrul** [comment]
That solution seems to be the same workaround in a different way. Most places that check for `has_con_24bit_colors()` also explicitly check for `!con_is_legacy`.

**Comment by dscho** [comment]
> 
> 
> That solution seems to be the same workaround in a different way. Most places that check for `has_con_24bit_colors()` also explicitly check for `!con_is_legacy`.

Interesting. Is there maybe a place where `!con_is_legacy` was missed?

**Comment by rimrul** [comment]
>Interesting. Is there maybe a place where `!con_is_legacy` was missed?

That sounds like it would manifest itself in the opposite way, i.e. Issues in legacy mode that don't exist in non-legacy mode, but I'll have a quick look.

**Comment by rimrul** [comment]
There are these:
P https://github.com/git-for-windows/msys2-runtime/blob/main/winsup/cygwin/fhandler_console.cc#L2987-L2988

https://github.com/git-for-windows/msys2-runtime/blob/main/winsup/cygwin/fhandler_console.cc#L1369-L1370

https://github.com/git-for-windows/msys2-runtime/blob/12e88192b84673b81c6550e48fb7df9b607dada3/winsup/cygwin/fhandler_console.cc#L1185-L1196

https://github.com/git-for-windows/msys2-runtime/blob/main/winsup/cygwin/fhandler_console.cc#L507-L509

https://github.com/git-for-windows/msys2-runtime/blob/main/winsup/cygwin/fhandler_console.cc#L517-L518

https://github.com/git-for-windows/msys2-runtime/blob/main/winsup/cygwin/fhandler_console.cc#L536-L537

https://github.com/git-for-windows/msys2-runtime/blob/main/winsup/cygwin/fhandler_console.cc#L585-L600

But they seem to be intentional.

**Comment by rimrul** [comment]
https://github.com/git-for-windows/msys2-runtime/blob/main/winsup/cygwin/fhandler_console.cc#L421-L434 seems like it could be related, though.

**Comment by dscho** [comment]
> 
> 
> https://github.com/git-for-windows/msys2-runtime/blob/main/winsup/cygwin/fhandler_console.cc#L421-L434 seems like it could be related, though.

You mean in https://github.com/git-for-windows/msys2-runtime/blob/12e88192b84673b81c6550e48fb7df9b607dada3/winsup/cygwin/fhandler_console.cc#L426, this diff might help things along?

```diff
-  /* Nothing to do for xterm compatible mode. */
-  if (wincap.has_con_24bit_colors () && !con_is_legacy)
-    return;
```

**Comment by rimrul** [comment]
It's just a hunch, but that early return in a function about setting cursor position does seem suspicious to me.

**Comment by sephiroth-j** [comment]
This is again a problem with v2.36.0 and Windows Terminal. In Git Bash, Vim uses `term=xterm` by default when used for rebase and the like. But when using Git with Windows Terminal without Git Bash, Vim uses `term=cygwin` there. I think it would be good to set `term=xterm-256color` or `term=xterm` globally in `%ProgramFiles%\Git\etc\vimrc`.

**Comment by dscho** [comment]
> This is again a problem with v2.36.0 and Windows Terminal. In Git Bash, Vim uses `term=xterm` by default when used for rebase and the like. But when using Git with Windows Terminal without Git Bash, Vim uses `term=cygwin` there. I think it would be good to set `term=xterm-256color` or `term=xterm` globally in `%ProgramFiles%\Git\etc\vimrc`.

@sephiroth-j could you test the latest [snapshot](https://wingit.blob.core.windows.net/files/index.html)? We changed the `TERM` value recently, which _might_ fix things for you.

**Comment by sephiroth-j** [comment]
> > This is again a problem with v2.36.0 and Windows Terminal. In Git Bash, Vim uses `term=xterm` by default when used for rebase and the like. But when using Git with Windows Terminal without Git Bash, Vim uses `term=cygwin` there. I think it would be good to set `term=xterm-256color` or `term=xterm` globally in `%ProgramFiles%\Git\etc\vimrc`.
> 
> @sephiroth-j could you test the latest [snapshot](https://wingit.blob.core.windows.net/files/index.html)? We changed the `TERM` value recently, which _might_ fix things for you.

@dscho, works out of the box with the snapshot from commit dc88e3cd72a2f0bbe2fe513acfc72bd66b577851. The changes from commit f8248abe963cb80a6aea260111f6f09748f4227d helped.

**Comment by dscho** [comment]
So I'll call this fixed in v2.36.1.

**Comment by TaleTN** [comment]
Yeah, it's fixed, w00t! Thanks. :)
